From d8e57a06a70bec4cd48002b6df7b9afa715c39a3 Mon Sep 17 00:00:00 2001
From: Nicholas Throm <narthrom@edu.uwaterloo.ca>
Date: Wed, 5 May 2021 00:36:56 -0400
Subject: [PATCH] Tweak reach lengths


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 21747b92..0b236e44 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1460,10 +1460,10 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         this.player.resetIdleTimer();
         if (entity != null) {
             boolean flag = this.player.hasLineOfSight(entity);
-            double d0 = 36.0D;
+            double d0 = 49.0D;
 
             if (!flag) {
-                d0 = 9.0D;
+                d0 = 36.0D;
             }
 
             if (this.player.h(entity) < d0) {
-- 
2.29.2

