From 28fb7a230d3775ff97c76eea06961920a911380e Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Mon, 28 Jun 2021 17:05:31 -0400
Subject: [PATCH] Revert "Fix null chunk provider crashes"

This reverts commit bc5f805907391f7fe236f128e134ea2588c64415.

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1d1123fd..1a92092b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -933,7 +933,7 @@ public final class CraftServer implements Server {
             }
 
             chunkProviderServer.chunkLoader = null;
-            // chunkProviderServer.chunkProvider = null; // This seems to cause crashes
+            chunkProviderServer.chunkProvider = null;
             chunkProviderServer.chunks.clear();
         }
 
@@ -1319,7 +1319,7 @@ public final class CraftServer implements Server {
 
         for (JsonListEntry entry : playerList.getProfileBans().getValues()) {
             result.add(getOfflinePlayer((GameProfile) entry.getKey()));
-        }
+        }        
 
         return result;
     }
-- 
2.29.2

