From 224dcd32b549ee763ba5526159aa7ff3bf8d4899 Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Tue, 14 Sep 2021 22:22:38 -0400
Subject: [PATCH] Fix null pointer error


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 7877a1cc..5fbc7b48 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -423,7 +423,11 @@ public class ChunkProviderServer implements IChunkProvider {
             }
         }
 
-        return this.chunkProvider.unloadChunks();
+        // SportPaper - Only unload if chunkProvider isn't null
+        if (this.chunkProvider != null) {
+            return this.chunkProvider.unloadChunks();
+        }
+        return true;
     }
 
     public boolean canSave() {
-- 
2.29.2

