From b65149594e9c403aada7d754f712d39c9957d4b3 Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Thu, 25 Nov 2021 15:48:02 -0500
Subject: [PATCH] Show more than 20 tps


diff --git a/src/main/java/org/spigotmc/TicksPerSecondCommand.java b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
index 21fd7efc..9de294ba 100644
--- a/src/main/java/org/spigotmc/TicksPerSecondCommand.java
+++ b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
@@ -38,6 +38,6 @@ public class TicksPerSecondCommand extends Command
     private static String format(double tps) // PaperSpigot - made static
     {
         return ( ( tps > 18.0 ) ? ChatColor.GREEN : ( tps > 16.0 ) ? ChatColor.YELLOW : ChatColor.RED ).toString()
-                + ( ( tps > 20.0 ) ? "*" : "" ) + Math.min( Math.round( tps * 100.0 ) / 100.0, 20.0 );
+                + ( ( tps > 20.0 ) ? "*" : "" ) + (Math.round( tps * 100.0 ) / 100.0);
     }
 }
-- 
2.29.2

