From f5f0feb849143c215ba880c928812d6b7d049325 Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Tue, 7 Dec 2021 16:24:24 -0500
Subject: [PATCH] Only reset inventory slots if required


diff --git a/src/main/java/net/minecraft/server/PlayerInventory.java b/src/main/java/net/minecraft/server/PlayerInventory.java
index 76fa51d9..75ff44e3 100644
--- a/src/main/java/net/minecraft/server/PlayerInventory.java
+++ b/src/main/java/net/minecraft/server/PlayerInventory.java
@@ -6,6 +6,7 @@ import java.util.concurrent.Callable;
 import java.util.List;
 
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.entity.HumanEntity;
 // CraftBukkit end
 
@@ -358,7 +359,11 @@ public class PlayerInventory implements IInventory {
             aitemstack = this.armor;
         }
 
-        aitemstack[i] = itemstack;
+        if (aitemstack[i] != null && CraftItemStack.asCraftMirror(itemstack).isSimilar(CraftItemStack.asCraftMirror(aitemstack[i]))) {
+            aitemstack[i].count = itemstack.count;
+        } else {
+            aitemstack[i] = itemstack;
+        }
     }
 
     public float a(Block block) {
-- 
2.29.2

