From dda9e7682d754bc5a8c0b51efa374d02114e0de4 Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Mon, 20 Jun 2022 21:30:59 -0400
Subject: [PATCH] Log ping spikes


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 160b9eb4..091ded22 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2095,6 +2095,9 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             if (this.player.ping == 0) {
                 this.player.ping = i;
             } else {
+                if (i * 100 / this.player.ping > 130) {
+                    System.out.println(this.player.displayName + " ping spiked from " + this.player.ping + "ms to " + i + "ms");
+                }
                 this.player.ping = (this.player.ping * 3 + i) / 4;
             }
         }
-- 
2.29.2

