From 413fa3e459edbe737730002d8b3886259dafbd52 Mon Sep 17 00:00:00 2001
From: halfmaster1 <ohpointfive@gmail.com>
Date: Thu, 23 Jun 2022 21:08:55 -0400
Subject: [PATCH] Fix block update NPE


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 1b667867..ac97438f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -149,7 +149,9 @@ public class CraftWorld implements World {
     }
 
     public Block getBlockAt(int x, int y, int z) {
-        return getChunkAt(x >> 4, z >> 4).getBlock(x & 0xF, y, z & 0xF);
+        Chunk chunk = getChunkAt(x >> 4, z >> 4);
+        if (chunk == null) { return null; }
+        return chunk.getBlock(x & 0xF, y, z & 0xF);
     }
 
     public int getBlockTypeIdAt(int x, int y, int z) {
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
index 7eaaf011..d9f9c5b0 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
@@ -165,6 +165,10 @@ public class CraftBlockState implements BlockState {
         requirePlaced();
         Block block = getBlock();
 
+        if (block == null) {
+            return false;
+        }
+
         if (block.getType() != getType()) {
             if (!force) {
                 return false;
@@ -293,4 +297,4 @@ public class CraftBlockState implements BlockState {
             throw new IllegalStateException("The blockState must be placed to call this method");
         }
     }
-}
\ No newline at end of file
+}
-- 
2.29.2

